Improve: Memory management for console hyperlinks

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Implements automatic cleanup of hyperlink data when links are no longer visible in console windows. When text containing links (created via `echoLink()`, `insertLink()`, or OSC 8 sequences) scrolls out of view or windows are cleared, the associated link data is now automatically removed.

#### Motivation for adding to Mudlet

While the existing link storage system worked correctly, it retained link data indefinitely even after the text was no longer visible. This change adopts a more efficient approach by tracking which links are currently referenced and cleaning up unreferenced data during normal buffer maintenance operations. This is particularly beneficial for long-running sessions with heavy link usage.

#### Other info (issues closed, discussion etc)

- Adds reference tracking to identify active links in buffer content
- Cleanup occurs automatically during `clearWindow()` and buffer shrink operations
- Includes comprehensive unit tests covering selective cleanup scenarios
- All existing tests pass
- Follows Qt6/C++20 best practices for memory management

mpconley committed:
Fix: TBuffer link store memory leak - implement reference-aware cleanup

- Add TLinkStore::removeUnreferencedLinks() for selective link cleanup
  * Only removes links that are no longer referenced by buffer content
  * Properly frees Lua references before removal to prevent registry leaks
  * Cleans up all link data: commands, hints, styling, expiry, selection groups

- Add TBuffer::collectActiveLinkIds() to scan buffer for referenced links
  * Iterates through all TChar objects to identify active link IDs
  * Uses QSet for O(1) lookup performance during cleanup

- Add TBuffer::clearLinkState() for comprehensive link cleanup
  * Calls removeUnreferencedLinks() with active link set
  * Cleans up link state tracking (hover/visited/selection states)
  * Resets interaction indices for removed links

- Trigger cleanup automatically:
  * TBuffer::clear() - after deleting all lines (removes all links)
  * TBuffer::shrinkBuffer() - after removing old lines (periodic cleanup)

- Add comprehensive unit tests in TLinkStoreTest:
  * testRemoveUnreferencedLinks() - validates selective removal
  * testRemoveUnreferencedLinksEmpty() - tests complete cleanup
  * testRemoveUnreferencedLinksNone() - tests preservation when all referenced

This prevents unbounded link store growth from repeated echoLink/insertLink/OSC 8
hyperlink usage, especially in long-running sessions or after clearWindow() calls.

Fixes memory accumulation in mLinkStore, mHintStore, mReferenceStore, mStylingStore,
mExpireStore, mExpireToLinks, and mSelectionGroupIndex that was not cleaned up by
the previous clearWindow() implementation.

SlySven reviewed (commented):
I've never really looked into links too closely but what happens when text containing a "link" is copied from the `TMainConsole` to another `TConsole` console - does the link information carry over into the new one and if so how are the references (ID numbers) handled - a new one for the new containing `TConsole`/`TBuffer` or is the same one used? I don't have the code in front of me at this time and I cannot remember whether the link storage is per-profile or per `TConsole`...

mpconley commented:
Good question! I investigated the link copying behavior to ensure this PR handles cross-console scenarios correctly.

When text with links is copied between consoles (via `appendBuffer()` or `appendFormatted()`), **new link IDs are created in the destination buffer's `TLinkStore`**. The link *data* (commands, hints, styling) is duplicated, but each console maintains completely independent link storage.

Link storage is per-`TBuffer` - each `TBuffer` has its own `mLinkStore` member (not per-profile). So when copying from `TMainConsole` to another `TConsole`:

1. The source link data is read from the source buffer's link store
2. New link entries are created in the destination buffer's link store with fresh IDs
3. The `TChar` objects in the destination reference these new IDs

This means the cleanup in this PR is safe. When `clearLinkState()` removes unreferenced links from a buffer during `clear()` or `shrinkBuffer()`, it only affects that specific buffer's link store and cannot impact links in other consoles.

The relevant code is in:
- [TBuffer.cpp:4392-4394](https://github.com/Mudlet/Mudlet/blob/f60d52e527b9c94faba928e81ca6e9377184fe36/src/TBuffer.cpp#L4392-L4394) (`appendBuffer`)
- [TBuffer.cpp:4120-4123](https://github.com/Mudlet/Mudlet/blob/f60d52e527b9c94faba928e81ca6e9377184fe36/src/TBuffer.cpp#L4120-L4123) (`appendFormatted`)

mpconley committed:
Merge branch 'development' into feature/link-memory-cleanup

vadi2 commented:
Nice work on the memory management for hyperlinks! The implementation looks solid. A few minor suggestions:

### 1. Add explicit `#include <QSet>` in TBuffer.h

The new `collectActiveLinkIds()` method returns `QSet<int>`, but `QSet` isn't directly included in `TBuffer.h`. While it works via transitive include from `TLinkStore.h`, explicit includes are more robust.

### 2. Add null-check for `mpHost` in `clearLinkState()`

The existing pattern in `TBuffer::updateColors()` logs a warning when `mpHost` is null. Consider the same here to help debug incomplete cleanup during shutdown:

```cpp
void TBuffer::clearLinkState()
{
    Host* pH = mpHost;
    if (!pH) {
        qWarning() << "TBuffer::clearLinkState() WARNING - mpHost is null, Lua references cannot be freed";
    }
    
    QSet<int> activeLinkIds = collectActiveLinkIds();
    mLinkStore.removeUnreferencedLinks(activeLinkIds, pH);
    // ...
}
```

### 3. Consider additional test cases

The current tests cover the basic scenarios well. Two edge cases worth testing:

- **Links with expire names**: Verify `mExpireStore` and `mExpireToLinks` are cleaned up properly
- **Links with Lua references**: Verify `mReferenceStore` entries are removed (even though `freeReference` is a no-op in test builds)

Example:
```cpp
void testRemoveUnreferencedLinksWithExpireNames()
{
    TLinkStore store(10);
    QStringList links, hints;
    links.append("command");
    hints.append("hint");
    
    int id1 = store.addLinks(links, hints, nullptr, QVector<int>(), "expire_group");
    int id2 = store.addLinks(links, hints);
    
    QSet<int> referencedIds;
    referencedIds.insert(id2);
    store.removeUnreferencedLinks(referencedIds, nullptr);
    
    QVERIFY(store.getExpireName(id1).isEmpty());
}
```

These are all minor - the core implementation is correct and well-structured.

mpconley committed:
Merge branch 'development' into feature/link-memory-cleanup

mpconley committed:
Improve: Address review feedback and code refinements

- Add explicit #include <QSet> in TBuffer.h
- Guard removeUnreferencedLinks call with mpHost null check
- Add comprehensive test coverage for removeUnreferencedLinks
- Extract removeLinkById helper to consolidate cleanup logic
- Enhance clearLinkState with mPendingSelectionStyling cleanup
- Add hint validation assertions to all test cases
- Use set intersection (&=) for cleaner pending styling cleanup
- Modernize iterator loop to use C++17 structured bindings
- Remove redundant mExpireToLinks.remove call

mpconley commented:
> A few minor suggestions

Implemented these suggestions, thanks!

vadi2 reviewed (approved):
Looks good... just need a pass on the obvious comments, for some reason they love to add them.

mpconley committed:
Merge branch 'development' into feature/link-memory-cleanup

mpconley committed:
Refine: Remove obvious comments that don't add value

Remove comments that simply restate what the code clearly shows:
- Comment about collecting link IDs (method name is clear)
- Comments about removing from stores (code is self-explanatory)
- Comments about identifying/removing links (obvious from code flow)

Condense verbose 3-line comment to concise single line explaining why
pending selection styling needs cleanup.
